x/staking/types/undelegate_authorization.go

```diff
@@ -0,0 +1,100 @@
+package types
+
+import (
+	"reflect"
+
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authz "github.com/cosmos/cosmos-sdk/x/authz/exported"
+)
+
+var (
+	_ authz.Authorization = &UndelegateAuthorization{}
+)
+
+// NewUndelegateAuthorization creates a new UndlegateAuthorization object.
+func NewUndelegateAuthorization(allowed []sdk.ValAddress, denied []sdk.ValAddress, amount *sdk.Coin) (*UndelegateAuthorization, error) {
```

Comment:
maybe we could try to refactor this and `NewDelegateAuthorization` so that there's less repeated code?